# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [4]:
def addTwo(x):
    return x+2

print(addTwo(4))

#lambda version
print((lambda x : x+2)(4))

anum = 4
print((lambda x:x+2)(anum))

6
6
6


#### Saving to a Variable

In [5]:
ftest = lambda x: x+2
ftest(4)

6

#### Multiple Inputs

In [7]:
print((lambda x, y, z :x * y * z)(3,5,8))

x_test = lambda x,y,z:x*y*z
print(x_test(3,5,8))



120
120


#### Passing a Lambda into a Function

In [1]:
def mul(f,num):
#     ...
#         f expects lambda function
#         num ecpects a number
#     ...
    return f(num)

mul(lambda x:x*x,4)

16

#### Returning a Lambda from a Function

In [10]:
def mul(num):
    return num *4

def returnFunc():
    test =4
    def mn(num):
        return num *2
    return mn

f_return =(returnFunc())
print(returnFunc())
print(f_return(4))


def returnlambda(b,c):
    return lambda x,a: x+a+b+c
rlamb=returnlambda(4,5)

print(rlamb(5,5))

<function returnFunc.<locals>.mn at 0x00000207401E5090>
8
19


#### If Statements within Lambdas

In [13]:
#lambda x : true if (condition) else false
fcondtion=lambda num:num*2 if num >10 else num +2

print(fcondtion(8))
print(fcondtion(12))
print(fcondtion(10))

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [14]:
f = lambda x: x**3
print(f(2))


8


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [19]:
#map(func, iterable)
#collections of data that can be iterable (list dict tuple set)
#normally the usage of map happens with a predefind function but we can use lambda as well
def squared(num, num2):
    if num <10 and num2<10:
        return num**2, num2 **2
    else:
        return   num,num2
        
numbers=[4,11,20,3,15,20]
more_nums=[4,10,3,2,6]

squared_nums_map = list(map(squared, numbers, more_nums))
print(squared_nums_map)


[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [21]:
#map(lambda x : do something, iterable)

#list(map(lambda x:x*2,nums))
squared_nums_lambda = list(map(lambda x,y:(x**2,y**2) if x<10 and y<10 else (x,y),numbers, more_nums))
print(squared_nums_lambda)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [26]:
print(list(map(lambda x,y:(x*2)-1,numbers, more_nums)))

[7, 21, 39, 5, 29]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [33]:
names = ['Bob','Andy','Wax','Evan','Angelica']

def a_name(name):
    if name[0].lower()=='a':
        return True
 
    else:
        return False
    
nnl=list(filter(a_name,names))

print(nnl)

['Andy', 'Angelica']


#### Using Lambda's with Filter()

In [34]:
newnameslambda= list(filter(lambda name:True if name[0].lower()=='a' else False,names))
print(newnameslambda)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [37]:
import statistics
numbers = [3,1,3,4,345,575,7,5,44,29]
mean = statistics.mean(numbers)
filtered_numbers = list(filter(lambda x: x>=mean, numbers))
print(filtered_numbers)

[345, 575]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [38]:
from functools import reduce

#reduce(function, iterable)

list_1=[2,4,6,8,10] 

def addNums(num1,num2):
    return num1 +num2

result_add =reduce(addNums, list_1)
print(result_add)

#the reduce does a function to the list of data

30


#### Using Lambda's with Reduce()

In [39]:
result= reduce(lambda x,y:x+y, list_1)
print(result)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [42]:
my_list=[1,2,3,4]

result = reduce(lambda x,y:x*y,my_list)

print(result)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [46]:
def addnums(num):
    if num<=1:
        print("addnums(1) =1")
        return num
    else:
        print(f"addnums({num})= {num}+ addnums({num -1})")
        return num+ addnums(num-1)
addnums(5)

addnums(5)= 5+ addnums(4)
addnums(4)= 4+ addnums(3)
addnums(3)= 3+ addnums(2)
addnums(2)= 2+ addnums(1)
addnums(1) =1


15

#### Writing a Factorial Function

In [48]:
def fact(num):
    if num <= 1:
        return 1
    else: 
        return num*fact(num-1)
    
fact(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [68]:
def subnum(num):
    if num== 0:
        return 0
    else:
        print(f"addnums({num})= {num}- addnums({num -1})")
        return num - subnum(num-1)

result=(subnum(5))
print(result)

addnums(5)= 5- addnums(4)
addnums(4)= 4- addnums(3)
addnums(3)= 3- addnums(2)
addnums(2)= 2- addnums(1)
addnums(1)= 1- addnums(0)
3


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [72]:
def my_range(stop, start, step =2):
    while start< stop:
        yield start
        start+=step
for i in my_range(20, start =2):
    my_generator_value = i
    print(my_generator_value)
    
my_range(20,start =2)

2
4
6
8
10
12
14
16
18


<generator object my_range at 0x00000207425A90E0>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [73]:
def squared_numbers_generator(n):
    while n > 0:
        yield n ** 2
        n -= 1
number = 5
squared_gen = squared_numbers_generator(number)
for squared_number in squared_gen:
    print(squared_number)

25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [74]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
filtered_places = list(filter(lambda x: x.strip() != "", places))
print(filtered_places)



['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [75]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda x: x.split()[-1].lower())

print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [76]:
# F = (9/5)*C + 32
places = [('Nashua', 32), ("Boston", 12), ("Los Angeles", 44), ("Miami", 29)]

converted_places = list(map(lambda x: (x[0], (9/5)*x[1] + 32), places))

print(converted_places)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [77]:
def fib(n, current=0, next_num=1, iteration=0):
    if iteration <= n:
        print(f"Iteration {iteration}: {current}")
        return fib(n, next_num, current + next_num, iteration + 1)

fib(5)

Iteration 0: 0
Iteration 1: 1
Iteration 2: 1
Iteration 3: 2
Iteration 4: 3
Iteration 5: 5
